dna/codon.go

```diff
@@ -158,107 +172,233 @@ func AminoAcidToString(a AminoAcid) string {
-//TODO: Add in frame as an argument.
-//TODO: Initialize to len 3 and use the frame argument to update the index instead of append.
-func BasesToCodons(b []Base) []*Codon {
-	var frame int
-	var answer []*Codon
-	var current []Base
-	for i := 0; i < len(b); i++ {
-		if frame == 2 {
-			current = append(current, b[i])
-			answer = append(answer, &Codon{Seq: current})
-			current = make([]Base, 0) //clear current
-			frame = 0
-		} else {
-			current = append(current, b[i])
-			frame++
-		}
+// OneLetterToAminoAcid converts a one letter amino acid byte into an AminoAcid type.
+func OneLetterToAminoAcid(b byte) AminoAcid {
+	switch b {
+	case 'A':
+		return Ala
+	case 'R':
+		return Arg
+	case 'N':
+		return Asn
+	case 'D':
+		return Asp
+	case 'C':
+		return Cys
+	case 'Q':
+		return Gln
+	case 'E':
+		return Glu
+	case 'G':
+		return Gly
+	case 'H':
+		return His
+	case 'I':
+		return Ile
+	case 'L':
+		return Leu
+	case 'K':
+		return Lys
+	case 'M':
+		return Met
+	case 'F':
+		return Phe
+	case 'P':
+		return Pro
+	case 'S':
+		return Ser
+	case 'T':
+		return Thr
+	case 'W':
+		return Trp
+	case 'Y':
+		return Tyr
+	case 'V':
+		return Val
+	case '*':
+		return Stop
+	default:
+		log.Panicf("cannot convert unrecognized byte %v to type AminoAcid", b)
+		return Stop
 	}
-	return answer
 }
 
-//CodonsToSeq reverts a slice of Codons into a slice of DNA bases.
-func CodonsToSeq(c []*Codon) []Base {
-	var answer []Base
-	for i := 0; i < len(c); i++ {
-		for j := 0; j < len(c[i].Seq); j++ {
-			answer = append(answer, c[i].Seq[j])
-		}
+// ThreeLetterToAminoAcid converts a three letter amino acid string into an AminoAcid type
+func ThreeLetterToAminoAcid(s string) AminoAcid {
+	switch s {
+	case "Ala":
+		return Ala
+	case "Arg":
+		return Arg
+	case "Asn":
+		return Asn
+	case "Asp":
+		return Asp
+	case "Cys":
+		return Cys
+	case "Gln":
+		return Gln
+	case "Glu":
+		return Glu
+	case "Gly":
+		return Gly
+	case "His":
+		return His
+	case "Ile":
+		return Ile
+	case "Leu":
+		return Leu
+	case "Lys":
+		return Lys
+	case "Met":
+		return Met
+	case "Phe":
+		return Phe
+	case "Pro":
+		return Pro
+	case "Ser":
+		return Ser
+	case "Thr":
+		return Thr
+	case "Trp":
+		return Trp
+	case "Tyr":
+		return Tyr
+	case "Val":
+		return Val
+	case "Ter":
+		return Stop
+	default:
+		log.Panicf("cannot convert unrecognized string %s into type AminoAcid", s)
+		return Stop
+	}
+}
+
+// BasesToCodons converts a slice of DNA bases into a slice of Codons.
+// Input expects bases to be in-frame. If the input sequence is not a
+// factor of three the function will panic.
+func BasesToCodons(b []Base) []Codon {
+	frame := len(b) % 3
+	answer := make([]Codon, 0, len(b)/3)
+	if !IsSeqOfACGT(b) {
+		log.Panicf("unrecognized base in sequence: %v, all input bases must be uppercase. N's are not allowed", b)
+	}
+	if frame != 0 {
+		log.Panicf("input sequence length is %d. input length must be a factor of 3", len(b))
+	}
+	for i := 0; i < len(b)-frame; i += 3 {
```

Comment:
Because we panic three lines above is frame !=0, maybe we can remove the subtraction of frame to simplify.

Reply:
Good point. This will be removed